src/tree/mod.rs

```diff
@@ -474,34 +498,84 @@ impl RatchetTree {
-            let (keypair, node_ciphertexts) = pair;
-            direct_path_nodes.push(DirectPathNode {
-                public_key: keypair.get_public_key().clone(),
+        for (public_key, node_ciphertexts) in public_keys.iter().zip(ciphertexts.iter()) {
+            direct_path_nodes.push(UpdatePathNode {
+                // TODO: don't clone ...
+                public_key: public_key.clone(),
                 encrypted_path_secret: node_ciphertexts.clone(),
             });
         }
-        DirectPath {
-            leaf_key_package,
-            nodes: direct_path_nodes,
-        }
+        Ok(direct_path_nodes)
     }
-    fn merge_public_keys(&mut self, direct_path: &DirectPath, path: Vec<NodeIndex>) {
-        assert_eq!(direct_path.nodes.len(), path.len()); // TODO return error
+
+    /// Merge public keys from a direct path to this tree along the given path.
+    fn merge_direct_path_keys(
+        &mut self,
+        direct_path: &UpdatePath,
+        path: Vec<NodeIndex>,
+    ) -> Result<(), TreeError> {
+        assert_eq!(direct_path.nodes.len(), path.len());
+        if direct_path.nodes.len() != path.len() {
+            return Err(TreeError::InvalidArguments);
+        }
+
         for (i, p) in path.iter().enumerate() {
             let public_key = direct_path.nodes[i].clone().public_key;
             let node = ParentNode::new(public_key.clone(), &[], &[]);
             self.nodes[p.as_usize()].node = Some(node);
         }
+
+        Ok(())
+    }
+
+    /// Merge public keys along the direct path of the own node.
+    fn merge_direct_path(
+        &mut self,
+        direct_path: &[NodeIndex],
+        keys: Vec<HPKEPublicKey>,
+    ) -> Result<(), TreeError> {
+        assert_eq!(direct_path.len(), keys.len());
+        if direct_path.len() != keys.len() {
+            return Err(TreeError::InvalidArguments);
+        }
+
+        for (key, path_index) in keys.iter().zip(direct_path) {
+            // TODO: this is pretty ugly
+            let node = match self.nodes.get(path_index.as_usize()) {
+                Some(node) => node.node.clone().unwrap(),
+                None => ParentNode::new(key.clone(), &[], &[]),
+            };
+            self.nodes[path_index.as_usize()].node = Some(node);
+        }
+
+        Ok(())
     }
-    pub fn merge_keypairs(&mut self, keypairs: &[HPKEKeyPair], path: &[NodeIndex]) {
-        assert_eq!(keypairs.len(), path.len()); // TODO return error
+
+    pub(crate) fn merge_public_keys(
+        &mut self,
+        public_keys: &[HPKEPublicKey],
+        path: &[NodeIndex],
+    ) -> Result<(), TreeError> {
+        assert_eq!(public_keys.len(), path.len());
+        if public_keys.len() != path.len() {
+            return Err(TreeError::InvalidArguments);
+        }
         for i in 0..path.len() {
-            let node = ParentNode::new(keypairs[i].get_public_key().clone(), &[], &[]);
+            // TODO: drop clone
```

Comment:
Can we drop the clone by using an iterator?

Reply:
This requires a couple more changes. Because we only have a reference here and the caller needs the keys again.
Ideally the caller should take them out of the tree after this action and this should consume. But we should do that in a follow up.